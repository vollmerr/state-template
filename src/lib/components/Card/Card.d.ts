import * as React from 'react';

export interface CardProps {
  /** Content to render */
  children?: React.ReactNode;

  /** Style class name to attach */
  className?: string;

  /** Footer to render */
  footer?: React.ReactNode;

  /** Header to render */
  header?: React.ReactNode;

  /** Action to clear the message */
  onDismiss?: () => void;

  /** Use style variant */
  variant?: 'default'
    | 'understated'
    | 'overstated'
    | 'standout'
    | 'primary'
    | 'danger'
    | 'inverted';
}

declare class Card extends React.Component<CardProps, {}> {}

export default Card;
