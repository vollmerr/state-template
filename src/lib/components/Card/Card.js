import React from 'react';
import classNames from 'classnames';
import T from 'prop-types';

import Icon from '../Icon';

const Card = (props) => {
  const {
    variant,
    header,
    footer,
    children,
    className,
    onDismiss,
    ...rest
  } = props;

  const cn = classNames([
    'card ',
    className,
    { [`card-${variant}`]: variant },
  ]);

  return (
    <div data-test={'card'} className={cn} {...rest}>
      {
        onDismiss
        && (
          <Icon
            name={'close-mark'}
            onClick={onDismiss}
            className={'dismiss-icon'}
            data-test={'button-dismiss'}
          />
        )
      }
      {header && <div className={'card-header'}>{header}</div>}
      <div className={'card-block'}>{children}</div>
      {footer && <div className={'card-footer'}>{footer}</div>}
    </div>
  );
};

Card.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    'default',
    'understated',
    'overstated',
    'primary',
    'danger',
    'inverted',
  ]),
  /** Header to render */
  header: T.node,
  /** Footer to render */
  footer: T.node,
  /** Content to render */
  children: T.node,
  /** Style class name to attach to button */
  className: T.string,
};

Card.defaultProps = {
  variant: 'default',
  header: null,
  footer: null,
  children: null,
  className: '',
};

export default Card;
