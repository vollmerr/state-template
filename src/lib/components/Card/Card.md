```jsx
  <Card header={'header goes here'} footer={'footer goes here'}>
    <p>content goes here...</p>
  </Card>
```

### variants

```jsx
  <Card>
    <p>default card</p>
  </Card>

  <Card variant={'understated'}>
    <p>understated card</p>
  </Card>

  <Card variant={'overstated'}>
    <p>overstated card</p>
  </Card>

  <Card variant={'standout'}>
    <p>standout card</p>
  </Card>

  <Card variant={'primary'}>
    <p>primary card</p>
  </Card>

  <Card variant={'danger'}>
      <p>danger card</p>
  </Card>

  <Card variant={'inverted'}>
    <p>inverted card</p>
  </Card>
```

### with header

```jsx  
const header = <p>footer goes here <Button href={'http://www.google.com'} text={'example button'} variant={'primary'}/></p>;

  <Card header={header}>
    <p>content goes here...</p>
  </Card>
```

### with footer

```jsx
  const footer = <p>footer goes here <Button href={'http://www.google.com'} text={'example button'} variant={'primary'}/></p>;

  <Card footer={footer}>
    <p>content goes here...</p>
  </Card>
```

