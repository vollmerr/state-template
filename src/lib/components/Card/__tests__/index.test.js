import React from 'react';
import { shallow } from 'enzyme';

import { Card as fromLib } from '../../..';
import Card from '..';

const TestHeader = () => <div>in header...</div>;
const TestChild = () => <div>test children...</div>;
const TestFooter = () => <div>in footer...</div>;

const props = {
  onDismiss: jest.fn(),
  header: <TestHeader />,
  children: <TestChild />,
  footer: <TestFooter />,
};

let wrapper;
describe('Card', () => {
  beforeEach(() => {
    wrapper = shallow(<Card {...props} />);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(Card);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should allow variants', () => {
    wrapper.setProps({ variant: 'overstated' });
    expect(wrapper.prop('className')).toMatch(/card-overstated/);
  });

  it('should attach classNames', () => {
    wrapper.setProps({ className: 'testClass' });
    expect(wrapper.prop('className')).toMatch(/testClass/);
  });

  it('should pass additional props', () => {
    wrapper.setProps({ id: 'testId' });
    expect(wrapper.prop('id')).toBe('testId');
  });

  it('should call `onDismiss` when the close icon is clicked', () => {
    wrapper.find({ 'data-test': 'button-dismiss' }).simulate('click');
    expect(props.onDismiss).toBeCalled();
  });
});
