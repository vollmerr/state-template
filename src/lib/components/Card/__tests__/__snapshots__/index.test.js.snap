// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly 1`] = `
<div
  className="card  card-default"
>
  <Icon
    className="dismiss-icon"
    data-test="button-dismiss"
    name="close-mark"
    onClick={[MockFunction]}
    tag="span"
  />
  <div
    className="card-header"
  >
    <TestHeader />
  </div>
  <div
    className="card-block"
  >
    <TestChild />
  </div>
  <div
    className="card-footer"
  >
    <TestFooter />
  </div>
</div>
`;
